File: d1/d1.py
import os

from collections import deque

def part1(input):
    res = 0
    for line in input.split('\n'):
        if not line: continue

        f = l = -1
        for c in line:
            if c.isnumeric():
                if f == -1:
                    f = c
                l = c
    
        res += int(str(f) + str(l))

    return res

def part2(input):
    res = 0
    
    """
    one, two, six, 
    four, five, nine
    three, seven, eight
    """

    nums = {
        'one': 1,
        'two': 2,
        'three': 3,
        'four': 4,
        'five': 5,
        'six': 6,
        'seven': 7,
        'eight': 8,
        'nine': 9
    }

    res = 0
    for line in input.split('\n'):
        if not line: continue

        f = l = -1
        for i, c in enumerate(line):
            if c.isnumeric():
                if f == -1:
                    f = c
                l = c
                continue

            t = -1
            x = ''.join(line[i - 2:i + 1])
            y = ''.join(line[i - 3:i + 1])
            z = ''.join(line[i - 4:i + 1])

            if x in ['one', 'two', 'six']:
                t = nums[x]
            elif y in ['four', 'five', 'nine']:
                t = nums[y]
            elif z in ['three', 'seven', 'eight']:
                t = nums[z]

            if t != -1:
                if f == -1:
                    f = t
                l = t

        res += int(str(f) + str(l))
    
    return res

if __name__ == "__main__":
    dir_path = os.path.dirname(os.path.realpath(__file__))
    i = open(dir_path + '\d1-in.txt', 'r').read()

    print(part1(i))
    print(part2(i))

File: d1/d1.cpp
#include <bits/stdc++.h>

using namespace std;

void part1(const vector<string>& lines) {
    int res;

    for (const string& line : lines) {
        char f = 0, l = 0;
        for (char c : line) {
            if (isdigit(c)) {
                if (f == 0) f = c;
                l = c;
            }
        }

        stringstream s;
        s << f << l;
        int z;
        s >> z;
        res += z;
    }

    cout << res << endl;
}

void part2(const vector<string>& lines) {
    int res;
    unordered_map<string, char> mp{
        {"one", '1'}, {"two", '2'},   {"three", '3'}, {"four", '4'}, {"five", '5'},
        {"six", '6'}, {"seven", '7'}, {"eight", '8'}, {"nine", '9'},
    };

    for (const string& line : lines) {
        char f = 0, l = 0;
        for (int i = 0; i < line.length(); i++) {
            if (isdigit(line[i])) {
                if (f == 0) f = line[i];
                l = line[i];
            } else {
                if (i >= 2 && mp.count(line.substr(i - 2, 3))) {
                    int t = mp.at(line.substr(i - 2, 3));
                    if (f == 0) f = t;
                    l = t;
                } else if (i >= 3 && mp.count(line.substr(i - 3, 4))) {
                    int t = mp.at(line.substr(i - 3, 4));
                    if (f == 0) f = t;
                    l = t;
                } else if (i >= 4 && mp.count(line.substr(i - 4, 5))) {
                    int t = mp.at(line.substr(i - 4, 5));
                    if (f == 0) f = t;
                    l = t;
                }
            }
        }

        stringstream s;
        s << f << l;
        int z;
        s >> z;
        res += z;
    }

    cout << res << endl;
}

int main() {
    ifstream inputFile("d1-in.txt");
    vector<string> lines;
    string line;

    if (inputFile.is_open()) {
        while (getline(inputFile, line)) {
            lines.push_back(line);
        }

        inputFile.close();
    }

    part2(lines);
    return 0;
}